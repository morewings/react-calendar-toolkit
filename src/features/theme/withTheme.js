import React, {useContext} from 'react';
import PropTypes from 'prop-types';

const ThemeContext = React.createContext({});

export const useThemeContext = () => useContext(ThemeContext);

export const Provider = ({children, theme}) => (
  <ThemeContext.Provider value={theme}>{children}</ThemeContext.Provider>
);

Provider.propTypes = {
  theme: PropTypes.shape({}),
  children: PropTypes.element.isRequired,
};
Provider.defaultProps = {
  theme: {},
};

export default WrappedComponent => ({theme, ...restProps}) => (
  <Provider theme={theme}>
    <WrappedComponent {...restProps} />
  </Provider>
);
