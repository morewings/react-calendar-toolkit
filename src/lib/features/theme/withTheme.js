import React, {useContext} from 'react';
import PropTypes from 'prop-types';

const ThemeContext = React.createContext({});

export const useThemeContext = () => useContext(ThemeContext);

export const Provider = ({children, theme}) => (
  <ThemeContext.Provider value={theme}>{children}</ThemeContext.Provider>
);

Provider.propTypes = {
  theme: PropTypes.shape({}),
  children: PropTypes.element.isRequired,
};
Provider.defaultProps = {
  theme: {},
};

// eslint-disable-next-line react/prop-types
const withTheme = WrappedComponent => ({theme, ...restProps}) => (
  <Provider theme={theme}>
    <WrappedComponent {...restProps} />
  </Provider>
);

export default withTheme;
