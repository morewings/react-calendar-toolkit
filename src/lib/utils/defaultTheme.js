const baseColor = '#43a047';
const gray = 'rgba(0, 0, 0, 0.38)';
const borderRadius = '4px';

const defaultTheme = {
  /** Set font for datepicker */
  fontPrimary: 'Helvetica Neue',
  /** Font, which is applied, if --fontPrimary is not available */
  fontSecondary: 'Arial',
  /** Fallback font if two previous are unavailable */
  fontFallback: 'sans-serif',
  fontSize: '14px',
  fontWeight: '400',
  /** Applied to the title */
  fontSizeSmall: '10px',
  /** Applied to the header text */
  fontSizeBig: '24px',
  /** Set inner width */
  calendarWidth: '240px',
  /** Calculate total width */
  datePickerWidth: 'calc(var(--calendarWidth) + var(--innerPadding) * 2)',
  selectorHeight: '36px',
  calendarHeight: '200px',
  /** Set padding between elements */
  innerPadding: '12px',
  /** Day, month, year entries height */
  entryHeight: '32px',
  entryMargin: '1px',
  datepickerBorderRadius: borderRadius,
  entryBorderRadius: borderRadius,
  /** Background color of Calendar */
  bgColor: 'white',
  /** Base color for text */
  textColor: 'black',
  headerBgColor: baseColor,
  weekendTextColor: 'red',
  headerTextColor: 'white',
  borderColor: gray,
  modalBgColor: gray,
  weekDayColor: gray,
  entryHoverColor: 'rgba(0, 0, 0, 0.08)',
  entrySelectedColor: baseColor,
  entryHighlightColor: 'lightblue',
  inputWidth: '90px',
  inputPadding: '8px',
  inputBorderRadius: borderRadius,
  inputBorderColor: gray,
  inputHighlightColor: 'lightblue',
};

export default defaultTheme;
