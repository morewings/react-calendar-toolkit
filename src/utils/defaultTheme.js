export default {
  '--dayWidth': '38px',
  '--dayHeight': 'var(--dayWidth)',
  '--basePadding': '6px',
  '--innerPadding': '16px',
  '--innerWidth': '280px',
  '--datePickerWidth': 'calc(var(--innerWidth) + var(--innerPadding) * 2)',
  '--borderRadius': '4px',
  '--gridHeight': '247px',
  '--inputHeight': '25px',
  '--baseColor': 'black',
  '--gray': 'rgba(0, 0, 0, 0.38)',
  '--lightGray': 'rgba(0, 0, 0, 0.08)',
  '--green': '#43a047',
  '--disabled': 'var(--gray)',
  '--otherMonthColor': 'var(--lightGray)',
  '--active': 'var(--green)',
  '--hoverColor': 'var(--lightGray)',
  '--headerBgColor': 'var(--green)',
  '--headerTextColor': 'white',
  '--weekendColor': 'red',
  '--highlightColor': 'lightblue',
};
