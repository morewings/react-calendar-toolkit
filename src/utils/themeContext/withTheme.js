import React from 'react';
import PropTypes from 'prop-types';
import ThemeContext from './Context';

const withTheme = WrappedComponent => {
  const Component = props => (
    <ThemeContext.Provider value={props.theme}>
      <WrappedComponent {...props} />
    </ThemeContext.Provider>
  );
  Component.propTypes = {
    theme: PropTypes.shape({}),
  };
  Component.defaultProps = {
    theme: {},
  };
  return Component;
};

export default withTheme;
