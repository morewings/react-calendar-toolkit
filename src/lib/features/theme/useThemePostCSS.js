import {useTheme} from 'css-vars-hook';
import defaults from 'lib/utils/defaultTheme';
import {useThemeContext} from './withTheme';
// import {removeCSSVariable, setCSSVariable} from './cssVariables';

/** @function
 * @name useThemePostCSS
 * @description React hook. Sets css variables from Context and defaultTheme
 * @param {Object.<string, string>} [defaultTheme=defaults] Default theme to override with context values
 * @return {[Object, Function]}
 */
const useThemePostCSS = (defaultTheme = defaults) => {
  const propsTheme = useThemeContext();
  const mergedTheme = {
    ...defaultTheme,
    ...propsTheme,
  };
  const {ref, setRef, style} = useTheme(mergedTheme);
  return {ref, setRef, style};
};

export default useThemePostCSS;
