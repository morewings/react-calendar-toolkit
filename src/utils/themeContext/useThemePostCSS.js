import {useRef, useCallback} from 'react';
import defaults from 'utils/defaultTheme';
import {useThemeContext} from './withTheme';
import {removeCSSVariable, setCSSVariable} from './cssVariables';

/** @function
 * @name useThemePostCSS
 * @description React hook. Sets css variables from Context and defaultTheme
 * @param {Object.<string, string>} [defaultTheme=defaults] Default theme to override with context values
 * @return {[Object, Function]}
 */
const useThemePostCSS = (defaultTheme = defaults) => {
  const ref = useRef(null);
  const propsTheme = useThemeContext();
  const mergedTheme = {
    ...defaultTheme,
    ...propsTheme,
  };
  const setRef = useCallback(
    element => {
      if (ref.current) {
        element &&
          Object.entries(mergedTheme).forEach(([variableName]) => {
            removeCSSVariable(element, variableName);
          });
      }
      if (element) {
        Object.entries(mergedTheme).forEach(([variableName, value]) => {
          setCSSVariable(element, variableName, value);
        });
      }
      ref.current = element;
    },
    [mergedTheme]
  );

  return [ref, setRef];
};

export default useThemePostCSS;
