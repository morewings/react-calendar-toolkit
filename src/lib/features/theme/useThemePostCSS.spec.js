import React from 'react';
import {render} from '@testing-library/react';
import {renderHook} from '@testing-library/react-hooks';
import defaultTheme from 'lib/utils/defaultTheme';
import useThemePostCSS from './useThemePostCSS';
import {Provider} from './withTheme';

const Component = ({themeOverride}) => {
  const {setRef, style} = useThemePostCSS(themeOverride);
  return (
    <div style={style} data-testid="styleContainer" ref={setRef}>
      Theme PostCSS
    </div>
  );
};

describe('useThemePostCSS', () => {
  it('sets default theme to provided HTMLElement', async () => {
    const {asFragment, getByTestId} = render(<Component />, {
      wrapper: ({children}) => <Provider>{children}</Provider>,
    });

    expect(asFragment()).toMatchSnapshot();

    Object.entries(defaultTheme).forEach(([key, value]) => {
      expect(
        getByTestId('styleContainer').style.getPropertyValue(`--${key}`)
      ).toBe(value);
    });
  });

  it('overrides default with custom theme when provided', async () => {
    const themeOverride = {
      foo: 'bar',
    };

    const {asFragment, getByTestId} = render(
      <Component themeOverride={themeOverride} />,
      {
        wrapper: ({children}) => <Provider>{children}</Provider>,
      }
    );

    expect(asFragment()).toMatchSnapshot();

    Object.entries(defaultTheme).forEach(([key]) => {
      expect(
        getByTestId('styleContainer').style.getPropertyValue(`--${key}`)
      ).toBe('');
    });

    Object.entries(themeOverride).forEach(([key, value]) => {
      expect(
        getByTestId('styleContainer').style.getPropertyValue(`--${key}`)
      ).toBe(value);
    });
  });

  it('overrides default with custom theme when provided', async () => {
    const themeOverride = {
      foo: 'bar',
      fiz: 'buz',
    };

    const propsOverride = {
      foo: 'baz',
    };

    const {asFragment, getByTestId} = render(
      <Component themeOverride={themeOverride} />,
      {
        wrapper: ({children}) => (
          <Provider theme={propsOverride}>{children}</Provider>
        ),
      }
    );

    expect(asFragment()).toMatchSnapshot();

    Object.entries(propsOverride).forEach(([key, value]) => {
      expect(
        getByTestId('styleContainer').style.getPropertyValue(`--${key}`)
      ).toBe(value);
    });
  });

  it('creates valid ref', () => {
    const element = document.createElement('div');
    const {result} = renderHook(() => useThemePostCSS(), {
      wrapper: ({children}) => <Provider>{children}</Provider>,
    });
    const {ref, setRef} = result.current;
    expect(result.current).toMatchSnapshot();
    setRef(element);
    expect(ref.current).toBe(element);
  });
});
