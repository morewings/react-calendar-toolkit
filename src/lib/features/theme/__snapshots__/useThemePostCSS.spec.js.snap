// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`useThemePostCSS creates valid ref 1`] = `
Object {
  "ref": Object {
    "current": null,
  },
  "setRef": [Function],
  "style": Object {
    "--bgColor": "white",
    "--borderColor": "rgba(0, 0, 0, 0.38)",
    "--calendarHeight": "200px",
    "--calendarWidth": "240px",
    "--datePickerWidth": "calc(var(--calendarWidth) + var(--innerPadding) * 2)",
    "--datepickerBorderRadius": "4px",
    "--entryBorderRadius": "4px",
    "--entryHeight": "32px",
    "--entryHighlightColor": "lightblue",
    "--entryHoverColor": "rgba(0, 0, 0, 0.08)",
    "--entryMargin": "1px",
    "--entrySelectedColor": "#43a047",
    "--fontFallback": "sans-serif",
    "--fontPrimary": "Helvetica Neue",
    "--fontSecondary": "Arial",
    "--fontSize": "14px",
    "--fontSizeBig": "24px",
    "--fontSizeSmall": "10px",
    "--fontWeight": "400",
    "--headerBgColor": "#43a047",
    "--headerTextColor": "white",
    "--innerPadding": "12px",
    "--inputBorderColor": "rgba(0, 0, 0, 0.38)",
    "--inputBorderRadius": "4px",
    "--inputHighlightColor": "lightblue",
    "--inputPadding": "8px",
    "--inputWidth": "90px",
    "--modalBgColor": "rgba(0, 0, 0, 0.38)",
    "--selectorHeight": "36px",
    "--textColor": "black",
    "--weekDayColor": "rgba(0, 0, 0, 0.38)",
    "--weekendTextColor": "red",
  },
}
`;

exports[`useThemePostCSS overrides default with custom theme when provided 1`] = `
<DocumentFragment>
  <div
    data-testid="styleContainer"
    style="--foo: bar;"
  >
    Theme PostCSS
  </div>
</DocumentFragment>
`;

exports[`useThemePostCSS overrides default with custom theme when provided 2`] = `
<DocumentFragment>
  <div
    data-testid="styleContainer"
    style="--foo: baz; --fiz: buz;"
  >
    Theme PostCSS
  </div>
</DocumentFragment>
`;

exports[`useThemePostCSS sets default theme to provided HTMLElement 1`] = `
<DocumentFragment>
  <div
    data-testid="styleContainer"
    style="--fontPrimary: Helvetica Neue; --fontSecondary: Arial; --fontFallback: sans-serif; --fontSize: 14px; --fontWeight: 400; --fontSizeSmall: 10px; --fontSizeBig: 24px; --calendarWidth: 240px; --datePickerWidth: calc(var(--calendarWidth) + var(--innerPadding) * 2); --selectorHeight: 36px; --calendarHeight: 200px; --innerPadding: 12px; --entryHeight: 32px; --entryMargin: 1px; --datepickerBorderRadius: 4px; --entryBorderRadius: 4px; --bgColor: white; --textColor: black; --headerBgColor: #43a047; --weekendTextColor: red; --headerTextColor: white; --borderColor: rgba(0, 0, 0, 0.38); --modalBgColor: rgba(0, 0, 0, 0.38); --weekDayColor: rgba(0, 0, 0, 0.38); --entryHoverColor: rgba(0, 0, 0, 0.08); --entrySelectedColor: #43a047; --entryHighlightColor: lightblue; --inputWidth: 90px; --inputPadding: 8px; --inputBorderRadius: 4px; --inputBorderColor: rgba(0, 0, 0, 0.38); --inputHighlightColor: lightblue;"
  >
    Theme PostCSS
  </div>
</DocumentFragment>
`;
